### PR TITLE
Refactor Changelog Window into class (#13765)

### PR DESCRIPTION
**Pull request by NBurley93**
I've refactored the changelog window into a Window class as according to issue #13765, I do feel like there can be improvements done on this implementation, but it is functional from the tests I've done on my machine.

### CODE REVIEW

File: src/openrct2-ui/windows/Changelog.cpp [hunk 1]
```diff
@@ -41,273 +41,300 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
-rct_window* window_changelog_open(int personality)
+class ChangelogWindow final : public Window
 {
-    rct_window* window;
-
-    window = window_bring_to_front_by_class(WC_CHANGELOG);
-    if (window != nullptr)
+private:
+    // Data
+
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Changelog.cpp:58]: You can remove those lines, class is private by default and we usually always have data at the top.

File: src/openrct2-ui/windows/Changelog.cpp [hunk 2]
```diff
@@ -41,273 +41,300 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
     }
-}
-
-static void window_new_version_process_info()
-{
-    _newVersionInfo = GetContext()->GetNewVersionInfo();
 
-    char version_info[256];
+private:
+    // Private funcs
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Changelog.cpp:227]: You can remove the obvious comment here.

File: src/openrct2-ui/windows/Changelog.cpp [hunk 3]
```diff
@@ -41,273 +41,300 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
+    if (window == nullptr)
+    {
+        // Create a new centered window
+        int32_t screenWidth = context_get_width();
+        int32_t screenHeight = context_get_height();
+        int32_t width = (screenWidth * 4) / 5;
+        int32_t height = (screenHeight * 4) / 5;
+
+        auto pos = ChangelogWindow::GetCentrePositionForNewWindow(width, height);
+        window = WindowCreate<ChangelogWindow>(WC_CHANGELOG, pos, width, height, WF_RESIZABLE);
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Changelog.cpp:336]: You can eliminate the static cast below by introducing a new variable here
```suggestion
        auto* newWindow = WindowCreate<ChangelogWindow>(WC_CHANGELOG, pos, width, height, WF_RESIZABLE);
```

File: src/openrct2-ui/windows/Changelog.cpp [hunk 4]
```diff
@@ -41,273 +41,300 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
+private:
+    // Data
+
+    const NewVersionInfo* _newVersionInfo;
+    std::vector<std::string> _changelogLines;
+    int32_t _changelogLongestLineWidth = 0;
+    int _persnality = 0;
+
+public:
+    // Public funcs
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Changelog.cpp:65]: Can be also removed

**Review by ZehMatt — CHANGES_REQUESTED** [review]
I left a few comments. I think some of the documentation might be a bit too verbose, I see no reason to include the return type or return values on booleans. The documentation should state what the function will do and what it will result in terms of what the value represents, this also avoids having to change comments when the types change. But I don't mind them too much if @OpenRCT2/developers are okay with it.

File: src/openrct2-ui/windows/Changelog.cpp [hunk 5]
```diff
@@ -41,273 +41,300 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
-
-    window = window_bring_to_front_by_class(WC_CHANGELOG);
-    if (window != nullptr)
+private:
+    // Data
+
+    const NewVersionInfo* _newVersionInfo;
+    std::vector<std::string> _changelogLines;
+    int32_t _changelogLongestLineWidth = 0;
+    int _persnality = 0;
```
Comment (IntelOrca) [inline src/openrct2-ui/windows/Changelog.cpp:62]:
```suggestion
    int personality = 0;
```

File: src/openrct2-ui/windows/Changelog.cpp [hunk 6]
```diff
@@ -41,273 +41,300 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
-    const int32_t lineHeight = font_get_line_height(FontSpriteBase::MEDIUM);
+    void OnMouseUp(rct_widgetindex widgetIndex) override
+    {
+        switch (widgetIndex)
+        {
+            case WIDX_CLOSE:
+                Close();
+                break;
+            case WIDX_OPEN_URL:
+                GetContext()->GetUiContext()->OpenURL(_newVersionInfo->url);
```
Comment (IntelOrca) [inline src/openrct2-ui/windows/Changelog.cpp:184]: I think we should null check `_newVersionInfo` where possible.

File: src/openrct2-ui/windows/Changelog.cpp [hunk 7]
```diff
@@ -41,273 +41,300 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
+    {
+        _newVersionInfo = GetContext()->GetNewVersionInfo();
 
-    _changelogLines.emplace_back(version_info);
-    _changelogLines.emplace_back("");
+        char version_info[256];
 
-    window_changelog_process_changelog_text(_newVersionInfo->changelog);
-}
+        const char* version_info_ptr = _newVersionInfo->name.c_str();
```
Comment (IntelOrca) [inline src/openrct2-ui/windows/Changelog.cpp:239]: Null check `_newVersionInfo`

**Comment by NBurley93** [comment]
Ok, I've made the changes suggested and made sure to rerun clang-format.

Follow-up question, are there plans to implement utility functions for window creation? I ended up needing to reimplement `GetCentrePositionForNewWindow` in ChangelogWindow in order to get the window to properly show in the center of the screen, which I don't believe to be an ideal solution if it can be helped.

File: src/openrct2-ui/windows/Changelog.cpp [hunk 8]
```diff
@@ -41,273 +41,308 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
-    window = window_bring_to_front_by_class(WC_CHANGELOG);
-    if (window != nullptr)
+    const NewVersionInfo* _newVersionInfo;
+    std::vector<std::string> _changelogLines;
+    int32_t _changelogLongestLineWidth = 0;
+    int _personality = 0;
+
+public:
+    /**
+     * @brief Retrieves a std::string from the changelog file
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Changelog.cpp:63]:
```suggestion
     * @brief Retrieves the changelog contents.
```

File: src/openrct2-ui/windows/Changelog.cpp [hunk 9]
```diff
@@ -41,273 +41,308 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
+        auto env = GetContext()->GetPlatformEnvironment();
+        return env->GetFilePath(PATHID::CHANGELOG);
     }
-    catch (const std::bad_alloc&)
+
+    /**
+     * @brief Attempts to read the changelog file, returns true on success
+     *
+     * @return true
+     * @return false
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Changelog.cpp:280]: Can be omitted

File: src/openrct2-ui/windows/Changelog.cpp [hunk 10]
```diff
@@ -41,273 +41,308 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
+            return false;
+        }
+
+        ProcessChangelogText(_changelogText);
+        return true;
     }
-    catch (const std::exception&)
+
+    /**
+     * @brief Pushes a line of text to the changelog
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Changelog.cpp:305]: Does not reflect what the function actually does.

File: src/openrct2-ui/windows/Changelog.cpp [hunk 11]
```diff
@@ -41,273 +41,308 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
+            _changelogLongestLineWidth + 4, (int32_t)_changelogLines.size() * font_get_line_height(FontSpriteBase::MEDIUM));
     }
-}
 
-static void window_changelog_process_changelog_text(const std::string& text)
-{
-    std::string::size_type pos = 0;
-    std::string::size_type prev = 0;
-    while ((pos = text.find("\n", prev)) != std::string::npos)
+    // TODO(NBurley93): This probably should be a utility function defined elsewhere for reusability
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Changelog.cpp:215]: TODO is fine but we don't put usernames in the code.

**Review by ZehMatt — CHANGES_REQUESTED** [review]

File: src/openrct2-ui/windows/Changelog.cpp (hunk 11 above)
Comment (duncanspumpkin) [reply src/openrct2-ui/windows/Changelog.cpp:215]: Technically in the coding style it says its fine to sign off comments with username. But ive not seen it in any.

Comment (NBurley93) [reply src/openrct2-ui/windows/Changelog.cpp:215]: I'll remove my username for brevity

Comment (ZehMatt) [reply src/openrct2-ui/windows/Changelog.cpp:215]: > Technically in the coding style it says its fine to sign off comments with username. But ive not seen it in any.

That is why I thought we don't do it, but given we have none perhaps we should change the style guide.

File: src/openrct2-ui/windows/Changelog.cpp (hunk 10 above)
Comment (NBurley93) [reply src/openrct2-ui/windows/Changelog.cpp]: I've updated this brief, please let me know if it's too verbose and I can cut it down a bit more.

**Comment by NBurley93** [comment]
More changes made, please let me know if the wording is a bit unclear or needs to be shortened for the doc comments.

File: src/openrct2-ui/windows/Changelog.cpp [hunk 12]
```diff
@@ -41,273 +41,306 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
+        min_width = MIN_WW;
+        min_height = MIN_WH;
+        max_width = MIN_WW;
+        max_height = MIN_WH;
+    }
 
-static void window_changelog_mouseup(rct_window* w, rct_widgetindex widgetIndex)
-{
-    switch (widgetIndex)
+    void OnClose() override
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Changelog.cpp:127]: Sorry I overlooked this part, you can remove this entirely including the DisposeChangelogData function, when the window is closed the destructor takes care of it.

**Review by ZehMatt — CHANGES_REQUESTED** [review]

File: src/openrct2-ui/windows/Changelog.cpp (hunk 12 above)
Comment (NBurley93) [reply src/openrct2-ui/windows/Changelog.cpp]: Done

File: src/openrct2-ui/windows/Changelog.cpp [hunk 13]
```diff
@@ -41,273 +41,293 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
 constexpr int32_t MIN_WW = 300;
 constexpr int32_t MIN_WH = 250;
 
-static rct_widget window_changelog_widgets[] = {
+static rct_widget s_wndChangelogWidgets[] = {
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/Changelog.cpp:44]: We don’t use Hungarian notation. `_windowChangelogWidgets` would be more fitting I think. @IntelOrca ?

Comment (NBurley93) [reply src/openrct2-ui/windows/Changelog.cpp]: I've made this change, do we want to make this change for the constexpr fields as well?

File: src/openrct2-ui/windows/Changelog.cpp [hunk 14]
```diff
@@ -41,273 +41,293 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CHANGELOG_TITLE;
...
+};
+
+rct_window* window_changelog_open(int personality)
+{
+    auto* window = window_bring_to_front_by_class(WC_CHANGELOG);
+    if (window == nullptr)
     {
-        log_error("Unable to read changelog.txt");
-        return false;
+        // Create a new centered window
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/Changelog.cpp:321]:
```suggestion
        // Create a new centred window
```

Very minor nitpick, but we used British English in the codebase.

**Review by Gymnasiast — APPROVED** [review]
All my points have been addressed. Leaving it to @ZehMatt or @IntelOrca to check their bits.

**Review by IntelOrca — APPROVED** [review]

**Pull request merged**